File: 270.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        double closest = root->val, min_diff = INT_MAX, diff;

        while(root != NULL) {
            diff = abs(root->val - target);
            if(diff < min_diff) {
                min_diff = diff;
                closest = root->val;
            } else if ((diff == min_diff) && (closest > root->val)) {
                closest = root->val;
            }
            if(root->val > target) {
                root = root->left;
            } else {
                root = root->right;
            }
        }
        return closest;
    }
};

File: 270.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        close = root.val
        min_diff = float('inf')

        while(root != None):
            diff = abs(root.val - target)
            if(diff < min_diff):
                min_diff = diff
                close = root.val
            elif(diff == min_diff and root.val != close):
                close = min(root.val, close)
            if(root.val > target):
                root = root.left
            else:
                root = root.right

        return close